src/StepParityGenerator.cpp

```diff
@@ -0,0 +1,491 @@
+#include "global.h"
+#include "StepParityGenerator.h"
+#include "StepParityCost.h"
+#include "NoteData.h"
+#include "TechCounts.h"
+#include "GameState.h"
+
+using namespace StepParity;
+
+void StepParityGenerator::analyzeNoteData(const NoteData &in)
+{	
+	columnCount = in.GetNumTracks();
+	
+	CreateRows(in);
+
+	if(rows.size() == 0)
+	{
+		LOG->Trace("StepParityGenerator::analyze no rows, bailing out");
+		return;
+	}
+	buildStateGraph();
+	analyzeGraph();
+}
+
+void StepParityGenerator::analyzeGraph() {
+	nodes_for_rows = computeCheapestPath();
+	ASSERT_M(nodes_for_rows.size() == rows.size(), "nodes_for_rows should be the same length as rows!");
+
+	for (unsigned long i = 0; i < rows.size(); i++)
+	{
+		StepParityNode *node = graph[nodes_for_rows[i]];
+		rows[i].setFootPlacement(node->state.columns);
+	}
+}
+
+void StepParityGenerator::buildStateGraph()
+{
+	// The first node of the graph is beginningState, which represents the time before
+	// the first note (and so it's roIndex is considered -1)
+	State beginningState(columnCount);
+	beginningState.rowIndex = -1;
+	beginningState.second = rows[0].second - 1;
+	StepParityNode *startNode = graph.addOrGetExistingNode(beginningState);
+
+	graph.startNode = startNode;
+
+	std::queue<State> previousStates;
+	previousStates.push(beginningState);
+	StepParityCost costCalculator(layout);
+
+	for (unsigned long i = 0; i < rows.size(); i++)
+	{
+		std::vector<State> uniqueStates;
+		Row &row = rows[i];
+		std::vector<FootPlacement> *PermuteFootPlacements = getFootPlacementPermutations(row);
+		while (!previousStates.empty())
+		{
+			State state = previousStates.front();
+			StepParityNode *initialNode = graph.addOrGetExistingNode(state);
+
+			for(auto it = PermuteFootPlacements->begin(); it != PermuteFootPlacements->end(); it++)
+			{
+				State resultState = initResultState(state, row, *it);
+				float* costs = costCalculator.getActionCost(&state, &resultState, rows, i);
+				resultState.calculateHashes();
+				StepParityNode *resultNode = graph.addOrGetExistingNode(resultState);
+				graph.addEdge(initialNode, resultNode, costs);
+				if(std::find(uniqueStates.begin(), uniqueStates.end(), resultState) == uniqueStates.end())
```

Comment:
There may be an opportunity to cut down on memory usage here. If I am reading this right at every stage of the search all unique nodes are being collected, and edges from all the previous nodes are created to the new set of unique nodes. Often the same state will be found from more than one edge from previous nodes. When that happens one of those paths will have a worse cost than the other. That worse edge can be removed. It also means that the previous node where that edge was coming from can be removed if it now has no edges leaving it to later nodes. That can be done recursively, pruning the tree.

At a high level what that is doing is saying that you have found two paths that both arrived at the same position at the same time on the pads. One of those paths must represent a better interpretation of the steps to get that position than the other. So all the data that is unique to that worse path is no longer needed because you know it is a worse interpretation that you shouldn't consider.

Similarly you may find nodes which have no valid paths out of them which satisfying the next state. In that case those nodes can similarly be pruned as they represent a dead end.

If you do this at every step of the search then when you get to the final row all you need to do is run this prune logic for every state in the current states that isn't the best current state. Then your tree is reduced to a linked list that represents the best path. The benefit to this approach is memory reduction, though. It may take slightly longer to run since at every row pruning occurs. So if you are interested in trying that approach I'd encourage some before/after tests to assess the performance.

For my chart parsing logic I considered this memory reduction to be the right tradeoff since charts can be extremely long and as I iterated on my logic I ended up increasing the number of kinds of steps to represent more and more moves. That means the number of edges out of each node started increasing, so it become more important to not let that memory get out of control. Also if in the future you parallelize this logic to run on multiple charts simultaneously you may have more of a need for less memory per thread. If you are interested in an example of this approach, [here is my prune logic](https://github.com/PerryAsleep/StepManiaLibrary/blob/0aa00e9d66c07f78eebf5d0685133d3e8661d1c1/StepManiaLibrary/ExpressedChart/ExpressedChart.cs#L1309).